### PR TITLE
[Filter/Lua] Add Lua tensor_filter subplugin

### PR DESCRIPTION
**Pull request by anyj0527**
- Add Lua filter working with Lua5.1
- For now, it only supports num_tensors == 1 for input/output and only supports uint8 type
- Add SSAT testcases
- Enable Lua filter in Tizen

TODO (in upcoming PR) :
- scale for num_tensors > 1
- Support various tensor type

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

### CODE REVIEW

File: ext/nnstreamer/tensor_filter/tensor_filter_lua.cc [hunk 1]
```diff
@@ -126,47 +215,121 @@ lua_subplugin::configure_instance (const GstTensorFilterProperties *prop)
...
-    /** @todo Error handling with load_stat */
+  /** @todo Support script given by raw string */
+  // script = loadScript (prop);
+  // int load_stat = luaL_loadbuffer (L, script, strlen (script), script);
+  // if (load_stat != 0) {
+  //   /** @todo Error handling with load_stat */
+  //  return;
+  // }
+
+  if (!g_file_test (prop->model_files[0], G_FILE_TEST_EXISTS)) {
```
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_filter/tensor_filter_lua.cc:233]: Discussion:
GstTensorFilterFrameworkInfo.verify_model_path 1 to check the file in tensor_filter main,
then remove this block.

Comment (anyj0527) [reply ext/nnstreamer/tensor_filter/tensor_filter_lua.cc]: When the lua script is given by model property as raw string, we should not take it as model path.

I just added the commit for supporting raw string script. If the given model prop value is invalid for file path, it takes it as lua script.

PTAL

File: ext/nnstreamer/tensor_filter/tensor_filter_lua.cc [hunk 2]
```diff
@@ -70,17 +74,97 @@ void _fini_filter_lua (void) __attribute__((destructor));
...
 
+/** @brief For getting value in Lua */
+static int
+tensor_index (lua_State *L)
+{
+  uint8_t** parray = (uint8_t **) luaL_checkudata (L, 1, "tensor");
+  int index = luaL_checkint (L, 2);
+  lua_pushnumber (L, (*parray)[index - 1]);
+
+  return 1;
```
Comment (dongju-chae) [inline ext/nnstreamer/tensor_filter/tensor_filter_lua.cc:85]: Q. Is there any meaning of this return value (e.g., 0 or 1)?

Comment (anyj0527) [reply ext/nnstreamer/tensor_filter/tensor_filter_lua.cc:85]: It means the number of values returning to lua.
With this C function, the lua script `some_tensor[index]` gets single value.

File: tests/nnstreamer_filter_lua/runTest.sh [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+PATH_TO_SCALER_MODEL="../test_models/models/scaler.lua"
+
+# Passthrough test
+gstTest "--gst-plugin-path=${PATH_TO_PLUGIN} videotestsrc num-buffers=1 ! video/x-raw,format=RGB,width=640,height=480,framerate=0/1 ! videoconvert ! video/x-raw, format=RGB ! tensor_converter ! tee name=t ! queue ! tensor_filter framework=lua model=\"${PATH_TO_PASSTHROUGH_MODEL}\" ! filesink location=\"testcase1.passthrough.log\" sync=true t. ! queue ! filesink location=\"testcase1.direct.log\" sync=true" 1 0 0 $PERFORMANCE
+callCompareTest testcase1.direct.log testcase1.passthrough.log 1 "Compare 1" 0 0
+
+# Passthrough -> Scaler test
+PATH_TO_SCRIPT="../test_models/models/scaler.py"
+ARGUMENTS="320x240"
+
```
Comment (gichan-jang) [inline tests/nnstreamer_filter_lua/runTest.sh:54]: Unnecessary lines.

Comment (anyj0527) [reply tests/nnstreamer_filter_lua/runTest.sh]: Thanks. I removed those lines

**Comment by myungjoo** [comment]
@semun-lee This is the LUA PR that I've told you a few days ago.

**Comment by myungjoo** [comment]
LGTM in overall.
After this PR:

1. a user manual for LUA script users.
2. in-line LUA scripting (and example)
3. negative unit tests (e.g., incorrect LUA script, inaccessible LUA script, incorrect pipeline for a given LUA script, ...)

**Comment by semun-lee** [comment]
> @semun-lee This is the LUA PR that I've told you a few days ago.

Thank you. We will try to use this feature in our distributed AI project.

File: ext/nnstreamer/tensor_filter/tensor_filter_lua.cc [hunk 4]
```diff
@@ -70,17 +74,97 @@ void _fini_filter_lua (void) __attribute__((destructor));
...
+}
+
+/** @brief For assigning new value in Lua */
+static int
+tensor_newindex (lua_State* L)
+{
+  uint8_t** parray = (uint8_t **) luaL_checkudata(L, 1, "tensor");
+  int index = luaL_checkint(L, 2);
+  int value = luaL_checkint(L, 3);
+  (*parray)[index - 1] = value;
```
Comment (zhoonit) [inline ext/nnstreamer/tensor_filter/tensor_filter_lua.cc:95]: is it okay to not check `index - 1` goes negative?

File: ext/nnstreamer/tensor_filter/tensor_filter_lua.cc [hunk 5]
```diff
@@ -70,17 +74,97 @@ void _fini_filter_lua (void) __attribute__((destructor));
...
+  static lua_subplugin *registeredRepresentation;
+  GstTensorsInfo inputInfo;
+  GstTensorsInfo outputInfo;
+
   lua_State *L;
   static const accl_hw hw_list[];
   static const int num_hw = 1;
 
+  const GstTensorMemory *input_for_lua;
+  GstTensorMemory *output_for_lua;
```
Comment (zhoonit) [inline ext/nnstreamer/tensor_filter/tensor_filter_lua.cc:166]: 👍

**Review by zhoonit — APPROVED** [review]
LGTM

File: ext/nnstreamer/tensor_filter/tensor_filter_lua.cc (hunk 4 above)
Comment (anyj0527) [reply ext/nnstreamer/tensor_filter/tensor_filter_lua.cc:95]: Thanks! I'm preparing the index checking in the following PR.

**Comment by myungjoo** [comment]
@jaeyun-jung Please merge this if you think it's appropriate.

**Review by myungjoo — APPROVED** [review]
LGTM

**Comment by jaeyun-jung** [comment]
@anyj0527 let's update build dependency in meson

**Comment by anyj0527** [comment]
@jaeyun-jung I updated the meson.build PTAL

**Review by jaeyun-jung — APPROVED** [review]

**Pull request merged**